### PR TITLE
Move HPC-gateway related SDK method to this repo

### PR DESCRIPTION
**Pull request by unkcpz**
The HPC-gateway app uses the available capabilities but with some modifications on the request and response format. The HPC-SDK https://github.com/unkcpz/hpc-sdk is designed to handle those requests over python. 
It will be a good idea to only have `python-sdk` to maintain everything.

### CODE REVIEW

File: marketplace/app/utils.py [hunk 1]
```diff
@@ -4,24 +4,32 @@
...
     Args:
         capability (str): capability that should be in capabilities
     """
 
-    @wraps(capability)
-    def wrapper(instance, *args, **kwargs):
-        if capability.__name__ not in instance.capabilities:
-            raise NotImplementedError("The app does not support this capability.")
-        return capability(instance, *args, **kwargs)
+    def decorator_check(func):
```
Comment (csadorf) [inline marketplace/app/utils.py:18]: You can implement an "optional" argument for a decorator like this:
```python
from functools import wraps, partial

def check(func, name=None):

    if callable(func):

        @wraps(func)
        def inner(*args, **kwargs):
            print(f"Running '{name or func.__name__}'.")
            func(*args, **kwargs)

        return inner

    elif isinstance(func, str):
        return partial(check, name=func)


@check
def foo():
    print("I am foo.")


@check("baz")
def bar():
    print("I am bar.")


foo()
bar()
```

File: marketplace/app/hpc_app.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            self.capabilities.append(camel_to_snake(capability["name"]))
+
+    @check_capability_availability
+    def heartbeat(self) -> str:
+        """Check the heartbeat of the application.
+
+        Returns:
+            str: heartbeat
+        """
+        return self.get(path="heartbeat").text
```
Comment (unkcpz) [inline marketplace/app/hpc_app.py:39]: This is a duplicate of `MarkepPlaceApp`. Any suggestion on how to organize them? @csadorf

File: .gitignore [hunk 3]
```diff
@@ -3,3 +3,5 @@
 __pycache__/
 build/
 dist/
+
+.env
```
Comment (csadorf) [inline .gitignore:7]: What is the motivation for this?

File: marketplace/app/hpc_app.py (hunk 2 above)
Comment (csadorf) [reply marketplace/app/hpc_app.py:39]: I'd suggest to make [`MarketPlaceApp`](https://github.com/materials-marketplace/python-sdk/blob/4e917e3f520647d26fead22aad2224bc2e79acf9/marketplace/app/marketplace_app.py#L16) inherit from this class and to remove this function.

File: marketplace/app/hpc_app.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        Actually it will create a new folder in specific user path on HPC
+        Return the folder name for further opreration."""
+        resp = self.post(path="newTransformation", json=config).text
+        resp = ast.literal_eval(resp)
+
+        return resp["resourceid"]
+
+    @check_capability_availability("get_transpormationList")
+    def list_job(self):
```
Comment (csadorf) [inline marketplace/app/hpc_app.py:84]:
```suggestion
    def list_jobs(self):
```

File: marketplace/app/hpc_app.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        """Check the heartbeat of the application.
+
+        Returns:
+            str: heartbeat
+        """
+        return self.get(path="heartbeat").text
+
+    @check_capability_availability("update_dataset")
+    def upload(self, resourceid, source_path=str):
+        """upload file to remote path `resourceid` from source path"""
```
Comment (csadorf) [inline marketplace/app/hpc_app.py:43]: Can you make sure to use consistent capitalization and formatting for your doc-strings?

File: marketplace/app/hpc_app.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            json={"filename": filename},
+        )
+        return resp.text
+
+    @check_capability_availability("new_transformation")
+    def new_job(self, config=None):
+        """Create a new job
+
+        Actually it will create a new folder in specific user path on HPC
+        Return the folder name for further opreration."""
```
Comment (csadorf) [inline marketplace/app/hpc_app.py:77]: This seems like an implementation detail to me.

File: marketplace/app/hpc_app.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    def list_job(self):
+        """List the jobs"""
+        return self.get(path="getTransformationList").json()
+
+    @check_capability_availability("startTransformation")
+    def run_job(self, resourceid):
+        """submit job in the path `resourceid`
+        It actually execute sbatch submit.sh in remote
+        Need job script `submit.sh` uploaded to the folder
+        TODO, check the job script ready"""
```
Comment (csadorf) [inline marketplace/app/hpc_app.py:93]: Can you remove all of these comments that appear developer-targeted?

File: marketplace/app/hpc_app.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        resp = self.post(
+            path="stopTransformation", params={"resourceid": f"{resourceid}"}
+        )
+
+        return resp.text
+
+    @check_capability_availability("delete_transformation")
+    def delete_job(self, resourceid):
+        """delete job corresponded to path `resourceid`
+        It actually drop entity from DB.
```
Comment (csadorf) [inline marketplace/app/hpc_app.py:112]: All of the "actually" comments should be removed.

File: marketplace/app/utils.py [hunk 9]
```diff
@@ -4,24 +4,36 @@
...
+        # get a func
+        func = func_or_str
+
+        @functools.wraps(func)
+        def wrapper(instance, *args, **kwargs):
+            _capability = capability or func.__name__
+
+            if _capability not in instance.capabilities:
+                print("capa: ", _capability)
+                print("list cap: ", instance.capabilities)
```
Comment (csadorf) [inline marketplace/app/utils.py:27]: left-over from debugging?

File: marketplace/app/utils.py [hunk 10]
```diff
@@ -4,24 +4,36 @@
 .. moduleauthor:: Pablo de Andres, Pranjali Singh (Fraunhofer IWM)
 """
 
+import functools
 import re
-from functools import wraps
 
 
-def check_capability_availability(capability):
+def check_capability_availability(func_or_str, capability=None):
```
Comment (csadorf) [inline marketplace/app/utils.py:11]:
```suggestion
def check_capability_availability(func_or_capability, capability=None):
```

File: marketplace/app/hpc_app.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        return resp.text
+
+    @check_capability_availability("delete_dataset")
+    def delete(self, resourceid, filename):
+        resp = self.delete(
+            path="deleteDataset",
+            params={"resourceid": f"{resourceid}"},
+            json={"filename": filename},
+        )
+        return resp.text
```
Comment (csadorf) [inline marketplace/app/hpc_app.py:70]: Why is this returning anything?

**Review by csadorf — CHANGES_REQUESTED** [review]
I assume that the function is fine, but this appears more like a draft to me. I'd appreciate if you could improve the code and documentation quality.

File: .gitignore (hunk 3 above)
Comment (unkcpz) [reply .gitignore:7]: I add an example `hpc_api.ipynb` which will read access_token from `.env`.

**Review by unkcpz — COMMENTED** [review]
@csadorf thanks! Updated upon your suggestion. And I have test with running it with HPC app.

**Review by csadorf — APPROVED** [review]
There are still a lot of issues with the doc-string formatting, however since this is not isolated to this change set, I think it is ok to merge and then work on improving that in a separate PR (#19).

**Comment by csadorf** [comment]
@pablo-de-andres Do you still want to review this or ok to merge?

**Comment by pablo-de-andres** [comment]
You can go ahead for me. I think I would maybe try to unify/simplify the behaviour of the decorator (to only check the function name, or to always require the capability), but it does not need to be done here.

**Pull request merged**